eth/downloader/resultstore.go

```diff
@@ -104,11 +114,41 @@ func (r *resultStore) GetDeliverySlot(headerNumber uint64) (*fetchResult, bool,
 	return res, stale, err
 }
 
+var headerSize uint64 = uint64(reflect.TypeOf(types.Header{}).Size())
+
+// isThrottled returns whether the given result index is throttled.
+func (r *resultStore) isThrottled(index int) bool {
+	return index >= r.throttleThreshold()
+}
+
+// throttleThreshold returns the index at which block requests will be throttled.
+// It is calculated by projecting an average worst-case expected block size for
+// in-flight block retrievals:
+//
+// Take the average gas used per in-flight block
+// and estimate the size of the block that could be constructed by filling the
+// block with a tx containing all-zero calldata (10 gas per byte of calldata).
+//
+// From this average worst-case block size, calculate the number of in-flight
+// retrievals that can be allowed until the cumulative block size hits a
+// predetermined threshold (1 gb).
+func (r *resultStore) throttleThreshold() int {
+	var avgGasUsed uint64
+	if r.itemsCount > 0 {
+		avgGasUsed = r.itemsGasUsed / uint64(r.itemsCount)
+	}
+	blockSize := max(estWorstCaseBlockSize(avgGasUsed), headerSize)
```

Comment:
I feel it’s a bit too strict to base the calculation on the worst-case scenario.

For example, a 15MGas block would be counted as 1.5 megabytes in the estimate, but in practice, on mainnet, the average block size over a short range is around 75.54 KiB. That’s about 20x smaller.

I think it would be reasonable to allow for some buffer. If the actual block size is, say, 5x larger than the estimated size, that should still be fine, especially since we already cap memory usage at 1GB. So I’d prefer to include an additional divisor (e.g., 5) when estimating block size to make the calculation more practical.

Reply:
> If the actual block size is, say, 5x larger than the estimated size, that should still be fine, especially since we already cap memory usage at 1GB

Instead of adding an arbitrary divisor, how about we up the cap to 5gb?  Effectively, that is the same thing here.